File: android/src/main/kotlin/com/valentingrigorean/arcgis_maps_flutter/authentication/ArcGISCredentialStoreController.kt
package com.valentingrigorean.arcgis_maps_flutter.authentication

import android.content.Context
import androidx.lifecycle.Lifecycle
import androidx.lifecycle.ViewModelProvider
import com.arcgismaps.ArcGISEnvironment
import com.arcgismaps.httpcore.authentication.ArcGISAuthenticationChallengeHandler
import com.arcgismaps.httpcore.authentication.ArcGISAuthenticationChallengeResponse
import com.arcgismaps.httpcore.authentication.ArcGISCredential
import com.arcgismaps.httpcore.authentication.ArcGISCredentialStore
import com.arcgismaps.httpcore.authentication.OAuthUserConfiguration
import com.arcgismaps.httpcore.authentication.OAuthUserCredential
import com.arcgismaps.httpcore.authentication.PregeneratedTokenCredential
import com.arcgismaps.httpcore.authentication.TokenCredential
import com.arcgismaps.httpcore.authentication.TokenInfo
import com.arcgismaps.mapping.PortalItem
import com.valentingrigorean.arcgis_maps_flutter.convert.authentication.toTokenInfoOrNull
import com.valentingrigorean.arcgis_maps_flutter.convert.toFlutterJson
import io.flutter.plugin.common.BinaryMessenger
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.launch
import java.time.Instant
import java.time.temporal.ChronoUnit
import java.util.UUID

class ArcGISCredentialStoreController(
    messenger: BinaryMessenger,
    private val scope: CoroutineScope,
    private val context: Context,
) :
    MethodChannel.MethodCallHandler {

    private val channel: MethodChannel =
        MethodChannel(messenger, "plugins.flutter.io/arcgis_channel/credential_store")
    private val credentials = mutableMapOf<String, ArcGISCredential>()

    private val oAuthConfigurations = mutableListOf<OAuthUserConfiguration>()

    private var store = ArcGISEnvironment.authenticationManager.arcGISCredentialStore

    init {
        channel.setMethodCallHandler(this)
    }

    fun dispose() {
        channel.setMethodCallHandler(null)
        setUpArcGISAuthenticationChallengeHandler()
    }

    override fun onMethodCall(call: MethodCall, result: MethodChannel.Result) {
        when (call.method) {
            "arcGISCredentialStore#makePersistent" -> {
                scope.launch {
                    ArcGISCredentialStore.createWithPersistence().onSuccess {
                        ArcGISEnvironment.authenticationManager.arcGISCredentialStore = it
                        store = it
                        result.success(null)
                    }.onFailure {
                        result.success(it.toFlutterJson())
                    }
                }
            }

            "arcGISCredentialStore#addCredential" -> {
                scope.launch {
                    val arguments = call.arguments as Map<*, *>
                    TokenCredential.create(
                        arguments["url"] as String,
                        arguments["username"] as String,
                        arguments["password"] as String,
                        arguments["tokenExpirationMinutes"] as Int?
                    ).onSuccess {
                        val uuid = UUID.randomUUID().toString()
                        credentials[uuid] = it
                        store.add(it)
                        result.success(uuid)
                    }.onFailure {
                        result.success(it.toFlutterJson())
                    }
                }
            }

            "arcGISCredentialStore#addPregeneratedTokenCredential" -> {
                val arguments = call.arguments as Map<*, *>
                val tokenInfo = arguments["tokenInfo"]!!.toTokenInfoOrNull()!!
                val token = PregeneratedTokenCredential(
                    arguments["url"] as String,
                    tokenInfo,
                    arguments["referer"] as String
                )
                store.add(token)
                val uuid = UUID.randomUUID().toString()
                credentials[uuid] = token
                result.success(uuid)
            }

            "arcGISCredentialStore#addOAuthCredential" -> {
                val arguments = call.arguments as Map<*, *>
                val oAuthConfiguration = OAuthUserConfiguration(
                    arguments["portalUrl"] as String,
                    arguments["clientId"] as String,
                    arguments["redirectUri"] as String,
                )
                oAuthConfigurations.add(oAuthConfiguration)
                scope.launch {
                    OAuthUserCredential.create(oAuthConfiguration) { oAuthUserSignIn ->
                        OAuthUserSignInActivity.promptForOAuthUserSignIn(context, oAuthUserSignIn)
                    }.onSuccess {
                        val uuid = UUID.randomUUID().toString()
                        credentials[uuid] = it
                        store.add(it)
                        result.success(uuid)
                    }.onFailure {
                        result.success(it.toFlutterJson())
                    }
                }
            }

            "arcGISCredentialStore.removeCredential" -> {
                val arguments = call.arguments as String
                val tokenCredential = credentials[arguments]
                if (tokenCredential != null) {
                    store.remove(tokenCredential)
                    credentials.remove(arguments)
                }
                result.success(null)
            }

            else -> result.notImplemented()
        }
    }


    private fun setUpArcGISAuthenticationChallengeHandler() {
        ArcGISEnvironment.authenticationManager.arcGISAuthenticationChallengeHandler =
            ArcGISAuthenticationChallengeHandler { challenge ->

                for (oAuthConfiguration in oAuthConfigurations) {
                    if (oAuthConfiguration.canBeUsedForUrl(challenge.requestUrl)) {
                        val oAuthUserCredential =
                            OAuthUserCredential.create(oAuthConfiguration) { oAuthUserSignIn ->
                                OAuthUserSignInActivity.promptForOAuthUserSignIn(
                                    context,
                                    oAuthUserSignIn
                                )
                            }.getOrThrow()

                        return@ArcGISAuthenticationChallengeHandler ArcGISAuthenticationChallengeResponse.ContinueWithCredential(
                            oAuthUserCredential
                        )
                    }
                }

                ArcGISAuthenticationChallengeResponse.ContinueAndFailWithError(
                    UnsupportedOperationException()
                )
            }
    }

}

File: ios/Classes/Authentication/ArcGISCredentialStoreController.swift
//
//  ArcGISCredentialStoreController.swift
//  arcgis_maps_flutter
//
//  Created by Valentin Grigorean on 06.07.2023.
//

import Foundation
import ArcGIS

extension TokenInfo {
    convenience init?(data: Dictionary<String, Any>) {
        let accessToken = data["accessToken"] as! String
        let expirationDate = data["expirationDate"] as! String
        let isSSLRequired = data["isSSLRequired"] as! Bool
        self.init(accessToken: accessToken, expirationDate: expirationDate.toDateFromIso8601()!, isSSLRequired: isSSLRequired)
    }
}

class ArcGISCredentialStoreController {
    private let taskManager = TaskManager()
    private let channel: FlutterMethodChannel
    private var credentialStore = ArcGISEnvironment.authenticationManager.arcGISCredentialStore
    private var tokenCredentials = [String: ArcGISCredential]()

    init(messenger: FlutterBinaryMessenger) {
        channel = UIFlutterMethodChannel(name: "plugins.flutter.io/arcgis_channel/credential_store", binaryMessenger: messenger)
        channel.setMethodCallHandler({ [weak self](call: FlutterMethodCall, result: @escaping FlutterResult) -> Void in
            guard let self else {
                return
            }
            self.handle(call: call, result: result)
        })

    }

    deinit {
        channel.setMethodCallHandler(nil)
    }

    private func handle(call: FlutterMethodCall,
                        result: @escaping FlutterResult) -> Void {
        switch (call.method) {
        case "arcGISCredentialStore#makePersistent":
            taskManager.createTask {
                let options = call.arguments as! [String: Any]
                let access = options["access"] as! Int
                let synchronizesWithiCloud = options["synchronizesWithiCloud"] as! Bool
                ArcGISEnvironment.authenticationManager.arcGISCredentialStore = try await .makePersistent(
                        access: self.getKeyAccess(value: access),
                        synchronizesWithiCloud: synchronizesWithiCloud
                )
                self.credentialStore = ArcGISEnvironment.authenticationManager.arcGISCredentialStore
                result(nil)
            }
            break
        case "arcGISCredentialStore#addCredential":
            taskManager.createTask {
                let data = call.arguments as! [String: Any]
                let url = data["url"] as! String
                let username = data["username"] as! String
                let password = data["password"] as! String
                let tokenExpirationMinutes = data["tokenExpirationMinutes"] as? Int

                do {
                    let tokenCredentials = try await TokenCredential.credential(
                            for: URL(string: url)!,
                            username: username,
                            password: password,
                            tokenExpirationMinutes: tokenExpirationMinutes
                    )
                    let uuid = UUID().uuidString
                    self.tokenCredentials[uuid] = tokenCredentials
                    self.credentialStore.add(tokenCredentials)
                    result(uuid)
                } catch {
                    result(error.toJSONFlutter())
                }
            }
            break
        case "arcGISCredentialStore#addPregeneratedTokenCredential":
            let data = call.arguments as! [String: Any]
            let url = data["url"] as! String
            let tokenInfo = TokenInfo(data: data["tokenInfo"] as! Dictionary<String, Any>)!
            let referer = data["referer"] as! String
            let token = PregeneratedTokenCredential(
                    url: URL(string: url)!,
                    tokenInfo: tokenInfo,
                    referer: referer
            )
            credentialStore.add(token)
            let uuid = UUID().uuidString
            tokenCredentials[uuid] = token
            result(uuid)
            break
        case "arcGISCredentialStore#addOAuthCredential":
            let data = call.arguments as! [String: Any]
            let portalUrl = data["portalUrl"] as! String
            let clientID = data["clientId"] as! String
            let redirectURL = data["redirectUri"] as! String
            let configuration = OAuthUserConfiguration(portalURL: URL(string: portalUrl)!, clientID: clientID, redirectURL: URL(string: redirectURL)!)
            taskManager.createTask{
                do {
                    let oAuthCredential = try await OAuthUserCredential.credential(for: configuration)
                    let uuid = UUID().uuidString
                    self.tokenCredentials[uuid] = oAuthCredential
                    self.credentialStore.add(oAuthCredential)
                    result(uuid)
                } catch {
                    result(error.toJSONFlutter())
                }
            }
            break
        case "arcGISCredentialStore#removeCredential":
            let uuid = call.arguments as! String
            let tokenCredential = tokenCredentials[uuid]
            if let tokenCredential = tokenCredential {
                credentialStore.remove(tokenCredential)
                tokenCredentials.removeValue(forKey: uuid)
            }
            result(nil)
            break
        default:
            result(FlutterMethodNotImplemented)
        }
    }

    private func getKeyAccess(value: Int) -> KeychainAccess {
        switch value {
        case 0:
            return .afterFirstUnlock
        case 1:
            return .afterFirstUnlockThisDeviceOnly
        case 2:
            return .whenUnlocked
        case 3:
            return .whenUnlockedThisDeviceOnly
        case 4:
            return .whenPasscodeSetThisDeviceOnly
        default:
            return .afterFirstUnlock
        }
    }
}
